File: frontend/src/api/exams.ts
/** Exams on the calendar. */

import type { Exam } from '../types'
import { request } from './client'

export function listExams(): Promise<Exam[]> {
  return request('/exams')
}

export function createExam(name: string, date: string, deckIds: string[]): Promise<Exam> {
  return request('/exams', { method: 'POST', body: JSON.stringify({ name, date, deck_ids: deckIds }) })
}

/** `deck_ids`, when present, replaces the exam's whole link set — the sheet always knows the full
 * selection, so there's no add/remove delta protocol to get out of sync. */
export function updateExam(
  id: string,
  patch: { name?: string; date?: string; deck_ids?: string[] },
): Promise<Exam> {
  return request(`/exams/${id}`, { method: 'PATCH', body: JSON.stringify(patch) })
}

export function deleteExam(id: string): Promise<void> {
  return request(`/exams/${id}`, { method: 'DELETE' })
}

File: backend/app/schemas/exams.py
"""Exams on the calendar, and the reference a deck tile carries to its next one."""

from __future__ import annotations

import uuid

# Aliased because these schemas have a *field* named `date`: the class-body assignment
# `date: date | None = None` would shadow the type in its own annotation (pydantic resolves
# annotation strings against the class namespace, where `date` is then the None default).
from datetime import date as Date

from pydantic import BaseModel


class ExamRef(BaseModel):
    """Just enough exam for a deck tile: the name for a tooltip, the date for a countdown."""

    name: str
    date: Date


class ExamOut(BaseModel):
    id: uuid.UUID
    name: str
    date: Date
    deck_ids: list[uuid.UUID]


class ExamCreate(BaseModel):
    name: str
    date: Date
    deck_ids: list[uuid.UUID] = []


class ExamUpdate(BaseModel):
    """`deck_ids` replaces the full link set when sent — the sheet UI always knows the whole
    selection, so set-replacement is both the simplest and the correct semantics."""

    name: str | None = None
    date: Date | None = None
    deck_ids: list[uuid.UUID] | None = None
